### ISSUE TITLE
csi: unpublish workflow ID mismatches

### ISSUE THREAD

**Issue by tgross** [hashicorp/nomad#7628]
Unfortunately the nightly e2e is still showing failures for CSI. The EFS test (which does not `-purge`) is working fine but the EBS test is failing:

```
05:25:11
  === RUN   TestE2E/CSI/*csi.CSIVolumesTest/TestEBSVolumeClaim
05:26:31      TestE2E/CSI/*csi.CSIVolumesTest/TestEBSVolumeClaim: utils.go:106:
05:26:31          	Error Trace:	utils.go:106
05:26:31          	            				csi.go:113
05:26:31          	Error:      	Condition never satisfied
05:26:31          	Test:       	TestE2E/CSI/*csi.CSIVolumesTest/TestEBSVolumeClaim
05:26:31      TestE2E/CSI/*csi.CSIVolumesTest/TestEBSVolumeClaim: utils.go:120:
05:26:31          	Error Trace:	utils.go:120
05:26:31          	            				csi.go:113
05:26:31          	Error:      	Should be true
05:26:31          	Test:       	TestE2E/CSI/*csi.CSIVolumesTest/TestEBSVolumeClaim
05:26:31          	Messages:   	allocations not placed for read-ebs-86520d54
05:26:31          	            	  blocked - created to place remaining allocations
05:26:31          	            	  complete -
```

I ran through the test by hand and put in some delays to see if I could get more info from the test failure. It looks like in the purge case we're seeing an `UpdateAlloc` that fails because the job is gone (as expected), but the claim is never being released.

The placement failure is:

```
Placement Failure
Task Group "group":
  * Constraint "CSI volume ebs-vol0 has exhausted its available writer claims": 2 nodes excluded by filter
```

Logs from the EBS server leader:

```
    2020-04-04T14:02:21.235Z [DEBUG] http: request complete: method=DELETE path=/v1/job/write-ebs-e1fe3604?purge=true duration=50.878564ms
    2020-04-04T14:02:22.612Z [DEBUG] http: request complete: method=GET path=/v1/jobs?prefix=write duration=237.377µs
    2020-04-04T14:02:24.040Z [DEBUG] http: request complete: method=GET path=/v1/agent/health?type=server duration=185.937µs
    2020-04-04T14:02:26.918Z [DEBUG] nomad.client: UpdateAlloc unable to find job: job=write-ebs-e1fe3604
    2020-04-04T14:02:51.299Z [DEBUG] worker: dequeued evaluation: eval_id=e08911bb-dd51-8ad7-18d8-b5697f294119
    2020-04-04T14:02:51.299Z [DEBUG] worker.service_sched: reconciled current state with desired state: eval_id=e08911bb-dd51-8ad7-18d8-b5697f294119 job_id=read-ebs-e1fe3604 namespace=default results="Total changes: (place 1) (destructive 0) (inplace 0) (stop 0)
Created Deployment: "3e75a6c5-2a12-527f-ce91-f5a61332ef2a"
Desired Changes for "group": (place 1) (inplace 0) (destructive 0) (stop 0) (migrate 0) (ignore 0) (canary 0)"
    2020-04-04T14:02:51.300Z [DEBUG] http: request complete: method=PUT path=/v1/jobs duration=8.917995ms
    2020-04-04T14:02:51.303Z [DEBUG] worker: created evaluation: eval="<Eval "631dab68-dc25-391d-f5bc-29f7d452d963" JobID: "read-ebs-e1fe3604" Namespace: "default">"
    2020-04-04T14:02:51.303Z [DEBUG] worker.service_sched: failed to place all allocations, blocked eval created: eval_id=e08911bb-dd51-8ad7-18d8-b5697f294119 job_id=read-ebs-e1fe3604 namespace=default blocked_eval_id=631dab68-dc25-391d-f5bc-29f7d452d963

```

Compare to logs from the successful EFS job:

```
    2020-04-04T14:04:38.454Z [DEBUG] http: request complete: method=DELETE path=/v1/job/write-efs-338d2569?purge=false duration=14.947127ms
    2020-04-04T14:04:38.509Z [DEBUG] http: request complete: method=PUT path=/v1/jobs duration=10.806967ms
    2020-04-04T14:04:38.509Z [DEBUG] worker: dequeued evaluation: eval_id=04696d9b-b1cd-d76f-9efb-caf6da918539
    2020-04-04T14:04:38.509Z [DEBUG] worker.service_sched: reconciled current state with desired state: eval_id=04696d9b-b1cd-d76f-9efb-caf6da918539 job_id=read-efs-338d2569 namespace=default results="Total changes: (place 1) (destructive 0) (inplace 0) (stop 0)
Created Deployment: "3db5bd96-e837-29df-a4ec-3defd2d6a4ff"
Desired Changes for "group": (place 1) (inplace 0) (destructive 0) (stop 0) (migrate 0) (ignore 0) (canary 0)"
    2020-04-04T14:04:38.517Z [DEBUG] worker: submitted plan for evaluation: eval_id=04696d9b-b1cd-d76f-9efb-caf6da918539
    2020-04-04T14:04:38.517Z [DEBUG] worker.service_sched: setting eval status: eval_id=04696d9b-b1cd-d76f-9efb-caf6da918539 job_id=read-efs-338d2569 namespace=default status=complete
```

On a whim I checked the _client_ logs, which suggest that the delete request is failing because the node plugin is unhappy about it:

```
    2020-04-04T14:02:21.190Z [WARN]  client.aws-ebs0: finished client unary call: grpc.code=Internal duration=2.022602ms grpc.service=csi.v1.Node grpc.method=NodeUnpublishVolume
    2020-04-04T14:02:21.227Z [WARN]  client.csi_client: finished client unary call: plugin.name=aws-ebs0 plugin.type=controller grpc.code=Internal duration=34.181929ms grpc.service=csi.v1.Controller grpc.method=ControllerUnpublishVolume
```

With that in mind, I tore down to get a clean cluster and retried, this time pausing to make sure I could get the plugin allocation logs so I can find out what the plugin should be telling our CSI client. (Yet another case of https://github.com/hashicorp/nomad/issues/7424)

```
I0404 14:32:04.128850       1 mount_linux.go:326] Disk successfully formatted (mkfs): ext4 - /dev/xvdba /csi/staging/ebs-vol0/rw-file-system-single-node-writer
I0404 14:32:04.128878       1 mount_linux.go:135] Mounting cmd (mount) with arguments ([-t ext4 -o defaults /dev/xvdba /csi/staging/ebs-vol0/rw-file-system-single-node-writer])
I0404 14:32:04.136025       1 node.go:143] NodeStageVolume: volume="ebs-vol0" operation finished
I0404 14:32:04.136087       1 node.go:145] donedone
I0404 14:32:04.137422       1 node.go:279] NodePublishVolume: called with args {VolumeId:vol-069c622e1bcb90890 PublishContext:map[devicePath:/dev/xvdba] StagingTargetPath:/csi/staging/ebs-vol0/rw-file-system-single-node-writer TargetPath:/csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/ebs-vol0/rw-file-system-single-node-writer VolumeCapability:mount:<> access_mode:<mode:SINGLE_NODE_WRITER >  Readonly:false Secrets:map[] VolumeContext:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 14:32:04.137465       1 node.go:440] NodePublishVolume: creating dir /csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/ebs-vol0/rw-file-system-single-node-writer
I0404 14:32:04.137479       1 node.go:450] NodePublishVolume: mounting /csi/staging/ebs-vol0/rw-file-system-single-node-writer at /csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/ebs-vol0/rw-file-system-single-node-writer with option [bind] as fstype ext4
I0404 14:32:04.137489       1 mount_linux.go:135] Mounting cmd (mount) with arguments ([-t ext4 -o bind /csi/staging/ebs-vol0/rw-file-system-single-node-writer /csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/ebs-vol0/rw-file-system-single-node-writer])
I0404 14:32:04.139574       1 mount_linux.go:135] Mounting cmd (mount) with arguments ([-t ext4 -o bind,remount /csi/staging/ebs-vol0/rw-file-system-single-node-writer /csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/ebs-vol0/rw-file-system-single-node-writer])
I0404 14:32:04.229522       1 node.go:349] NodeGetCapabilities: called with args {XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}

...

I0404 14:32:36.667566       1 node.go:324] NodeUnpublishVolume: called with args {VolumeId:vol-069c622e1bcb90890 TargetPath:/csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/vol-069c622e1bcb90890/rw-file-system-single-node-writer XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 14:32:36.667592       1 node.go:335] NodeUnpublishVolume: unmounting /csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/vol-069c622e1bcb90890/rw-file-system-single-node-writer
I0404 14:32:36.667599       1 mount_linux.go:202] Unmounting /csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/vol-069c622e1bcb90890/rw-file-system-single-node-writer
E0404 14:32:36.669154       1 driver.go:109] GRPC error: rpc error: code = Internal desc = Could not unmount "/csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/vol-069c622e1bcb90890/rw-file-system-single-node-writer": unmount failed: exit status 32
Unmounting arguments: /csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/vol-069c622e1bcb90890/rw-file-system-single-node-writer
Output: umount: /csi/per-alloc/dc1a3f3d-50ad-4bfe-b0b1-cbebe9a69d8d/vol-069c622e1bcb90890/rw-file-system-single-node-writer: no mount point specified.
I0404 14:32:36.669652       1 node.go:204] NodeUnstageVolume: called with args {VolumeId:vol-069c622e1bcb90890 StagingTargetPath:/csi/staging/vol-069c622e1bcb90890/rw-file-system-single-node-writer XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 14:32:36.669832       1 node.go:228] NodeUnstageVolume: /csi/staging/vol-069c622e1bcb90890/rw-file-system-single-node-writer target not mounted
```

The publish and unpublish workflows are operating on _different paths_. The publish side is mounting using the `VolumeID`, whereas the unpublish side is using the `ExternalID`. So this isn't a difference between using `-purge` and not using `-purge`, but a difference in behavior of the EBS vs EFS plugin test. We never see the RPC calls in the EFS plugin because they're it doesn't log them, but we're returning ok there because the node unpublish tolerates some unmount errors because we have to account for the mount path being cleaned up by the allocdir hook. The EFS plugin doesn't appear to be firing the `NodeUnstage` that would be unrecoverable, just the `NodeUnpublish`.

The reason the unpublish workflow at the client is is using `ExternalID` is because we called `RemoteID` in the GC instead of just using the volume's ID. Which isn't what we want for the node workflow, only for the controller workflow.

https://github.com/hashicorp/nomad/blob/98cbe43c9baa8d0fb01ebbb48b0c930bb12ed1b4/nomad/core_sched.go#L865
https://github.com/hashicorp/nomad/blob/98cbe43c9baa8d0fb01ebbb48b0c930bb12ed1b4/nomad/structs/csi.go#L267-L272

**Comment by tgross** [comment]
After running e2e again with this change, it looks like controller detach is also using an incorrect ID, but this time for the instance ID:

```
I0404 15:29:02.601864       1 cloud.go:364] AttachVolume volume="vol-031ed19c8bafc5559" instance="i-028eb38e5d0139d0a" request returned {
  AttachTime: 2020-04-04 15:29:02.578 +0000 UTC,
  Device: "/dev/xvdba",
  InstanceId: "i-028eb38e5d0139d0a",
  State: "attaching",
  VolumeId: "vol-031ed19c8bafc5559"
}
I0404 15:29:14.840211       1 manager.go:197] Releasing in-process attachment entry: /dev/xvdba -> volume vol-031ed19c8bafc5559
I0404 15:29:14.840237       1 controller.go:268] ControllerPublishVolume: volume vol-031ed19c8bafc5559 attached to node i-028eb38e5d0139d0a through device /dev/xvdba
I0404 15:29:18.328348       1 controller.go:298] ControllerGetCapabilities: called with args {XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 15:29:21.515718       1 controller.go:275] ControllerUnpublishVolume: called with args {VolumeId:vol-031ed19c8bafc5559 NodeId:ff2bb252-9c1a-53d3-e164-0388b1b4f881 Secrets:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
E0404 15:29:21.554203       1 driver.go:109] GRPC error: rpc error: code = Internal desc = Could not detach volume "vol-031ed19c8bafc5559" from node "ff2bb252-9c1a-53d3-e164-0388b1b4f881": error listing AWS instances: "InvalidInstanceID.Malformed: Invalid id: \"ff2bb252-9c1a-53d3-e164-0388b1b4f881\"\n\tstatus code: 400, request id: 464312e3-3060-490d-80ce-ea57568281ba"
I0404 15:29:48.329830       1 controller.go:298] ControllerGetCapabilities: called with args {XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 15:30:18.330861       1 controller.go:298] ControllerGetCapabilities: called with args {XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
```

Looks like we dropped the logic starting here: https://github.com/hashicorp/nomad/pull/7596/files#diff-4b1db1f4d39504d4eb463f858eeb57f0L584

**Comment by tgross** [comment]
Run with that build, the node plugin returns an error but that's not getting up to the caller, which then times out when we try to do the controller unpublish.

Node plugin:
```
I0404 16:28:08.918594       1 node.go:324] NodeUnpublishVolume: called with args {VolumeId:ebs-vol0 TargetPath:/csi/per-alloc/897986de-1835-1ca7-0fe7-b5fff8615d7f/ebs-vol0/rw-file-system-single-node-writer XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 16:28:08.918651       1 node.go:335] NodeUnpublishVolume: unmounting /csi/per-alloc/897986de-1835-1ca7-0fe7-b5fff8615d7f/ebs-vol0/rw-file-system-single-node-writer
I0404 16:28:08.918662       1 mount_linux.go:202] Unmounting /csi/per-alloc/897986de-1835-1ca7-0fe7-b5fff8615d7f/ebs-vol0/rw-file-system-single-node-writer
E0404 16:28:08.920179       1 driver.go:109] GRPC error: rpc error: code = Internal desc = Could not unmount "/csi/per-alloc/897986de-1835-1ca7-0fe7-b5fff8615d7f/ebs-vol0/rw-file-system-single-node-writer": unmount failed: exit status 32
Unmounting arguments: /csi/per-alloc/897986de-1835-1ca7-0fe7-b5fff8615d7f/ebs-vol0/rw-file-system-single-node-writer
Output: umount: /csi/per-alloc/897986de-1835-1ca7-0fe7-b5fff8615d7f/ebs-vol0/rw-file-system-single-node-writer: no mount point specified.
I0404 16:28:08.920552       1 node.go:204] NodeUnstageVolume: called with args {VolumeId:ebs-vol0 StagingTargetPath:/csi/staging/ebs-vol0/rw-file-system-single-node-writer XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 16:28:08.920776       1 node.go:228] NodeUnstageVolume: /csi/staging/ebs-vol0/rw-file-system-single-node-writer target not mounted
```

Controller plugin:
```
I0404 16:23:38.483332       1 controller.go:275] ControllerUnpublishVolume: called with args {VolumeId:vol-06dc899cb1e3d41b5 NodeId:i-02426e913269bd3f5 Secrets:map[] XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 16:24:06.015998       1 controller.go:298] ControllerGetCapabilities: called with args {XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 16:24:36.017104       1 controller.go:298] ControllerGetCapabilities: called with args {XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 16:25:06.018176       1 controller.go:298] ControllerGetCapabilities: called with args {XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
I0404 16:25:36.018994       1 controller.go:298] ControllerGetCapabilities: called with args {XXX_NoUnkeyedLiteral:{} XXX_unrecognized:[] XXX_sizecache:0}
E0404 16:25:55.963634       1 driver.go:109] GRPC error: rpc error: code = Internal desc = Could not detach volume "vol-06dc899cb1e3d41b5" from node "i-02426e913269bd3f5": RequestCanceled: request context canceled
caused by: context canceled
```

Looks like the 2 minute timeout was not enough here.

**Comment by tgross** [comment]
After fixing one more issue with the external vs internal ID, I'm seeing a weird behavior on EBS. The controller unpublish RPC is fired but times out waiting for the volume to become detached from the EC2 instance. And then immediately after the timeout is hit, the EBS volume is detached! 

In theory the AWS EBS plugin will wait up to [~38min](https://github.com/kubernetes-sigs/aws-ebs-csi-driver/blob/f55767a35e0cdc98976b1a175489160aaa5e84b9/pkg/cloud/cloud.go#L425) for the volume to become detached, but this suggests something is preventing the detachment until after we give up on it.

**Comment by tgross** [comment]
In the [Kubernetes implementation](https://github.com/kubernetes/kubernetes/blob/47f5d2923f3f35adc66b4797a95720f67c948b4e/pkg/volume/csi/csi_attacher.go#L397-L408) they're not waiting for the detachment process to complete from the plugin, but polling afterwards. That suggests that there may be a reason for this.

**Comment by tgross** [comment]
I'm going to open a new issue for the controller timeout problem. I think this is more of a workflow issue with allocation stops via job deregister.

**Comment by tgross** [comment]
Closed by #7632
